nixos/modules/services/databases/postgresql.nix

```diff
@@ -36,6 +36,32 @@ let
 
   groupAccessAvailable = versionAtLeast postgresql.version "11.0";
 
+  upgradeScript =
+    assert lib.versionOlder cfg.upgradeFrom.package.version cfg.package.version;
+    pkgs.writeShellApplication {
+      name = "upgrade.sh";
+      runtimeInputs = [];
+      text = ''
+        # There is no old version to upgrade from - noop
+        [[ -d "${cfg.upgradeFrom.dataDir}" ]] || exit 0
+
+        if [[ -f "${cfg.upgradeFrom.dataDir}/.upgraded" ]]; then
+          echo "Old data dir found, but was already migrated. Please remove the old path '${cfg.upgradeFrom.dataDir}' to suppress this message." >&2
+          exit 0
+        fi
+
+        pushd "${cfg.upgradeFrom.dataDir}"
+        ${cfg.package}/bin/pg_upgrade ${lib.cli.toGNUCommandLineShell {} {
+          old-datadir = cfg.upgradeFrom.dataDir;
+          new-datadir = cfg.dataDir;
+          old-bindir = "${cfg.upgradeFrom.package}/bin";
+          new-bindir = "${cfg.package}/bin";
+        }}
```

Comment:
There is no way to supply --link here which significantly speeds things up. Also other settings cannot be set. We should make a freeform setting type to supply more things and set the existing parameters per default.

Reply:
Ok I added an `extraFlags` option but I'm not sure it entirely matches https://www.postgresql.org/docs/current/pgupgrade.html

What do you think?

Reply:
hm. Let me refactor these all into just one `flags` option.

Reply:
Done.